File: android/app/src/main/java/io/mosip/greetings/cryptography/SenderTransfersOwnershipOfData.java
package io.mosip.greetings.cryptography;

/*
vk: verifier key
wk: wallet key

The sender transfers the ownership of the data, so the sender should use the receivers key to encrypt
wallet sending data to verifier | w->v: wallet encrypts with vk and verifier decrypts with vk
verifier sending data to wallet | v->w: verifier encrypts with wk and wallet decrypts with wk
 */
class SenderTransfersOwnershipOfData implements SecretsTranslator {
    private byte[] initVector;
    private CipherBox senderCryptoBox;
    private CipherBox receiverCryptoBox;

    public SenderTransfersOwnershipOfData(byte[] initVector, CipherBoxPackage cipherBoxPackage) {
        this.initVector = initVector;
        this.senderCryptoBox = cipherBoxPackage.getSelf();
        this.receiverCryptoBox = cipherBoxPackage.getOther();
    }

    @Override
    public byte[] initializationVector() {
        return initVector;
    }

    @Override
    public byte[] encryptToSend(byte[] plainText) {
        byte[] encrypt = receiverCryptoBox.encrypt(plainText);
        return encrypt;
    }

    @Override
    public byte[] decryptUponReceive(byte[] cipherText) {
        byte[] plainText = senderCryptoBox.decrypt(cipherText);
        return plainText;
    }
}

File: ios/bluetoothdemo/bluetoothdemo/cryptography/SecretTranslator/SenderTransfersOwnershipOfData.swift
//
//  SenderTransfersOwnershipOfData.swift
//  bluetoothdemo
//
//  Created by ShreeThaanu on 16/12/22.
//

import Foundation

class SenderTransfersOwnershipOfData: SecretsTranslator {

    var senderCipherBox: CipherBox
    var receiverCipherBox: CipherBox
    var initVector: Data

    init(cipherPackage: CipherPackage, initVector: Data) {
        self.senderCipherBox = cipherPackage.getSelfCipherBox
        self.receiverCipherBox = cipherPackage.getOtherCipherBox
        self.initVector = initVector
        self.senderCipherBox.printSecretKey(identifieris: "SENDER")
        self.receiverCipherBox.printSecretKey(identifieris: "RECIVER")
    }

    func initializationVector() -> Data {
        return initVector
    }

    func encryptToSend(data: Data) -> Data {
        let encrypt = (receiverCipherBox.encrypt(message: data))
        return encrypt
    }

    func decryptUponReceive(data: Data) -> Data {
        let encrypt = (senderCipherBox.decrypt(message: data))
        return encrypt
        
    }
}
